client/src/main/java/io/pravega/client/stream/impl/TransactionalEventStreamWriterImpl.java

```diff
@@ -0,0 +1,217 @@
+
+import static io.pravega.common.concurrent.Futures.getAndHandleExceptions;
+
+/**
+ * This class takes in events, finds out which segment they belong to and then calls write on the appropriate segment.
+ * It deals with segments that are sealed by re-sending the unacked events to the new correct segment.
+ * 
+ * @param <Type> The type of event that is sent
+ */
+@Slf4j
+@ToString(of = { "stream", "closed" })
+public class TransactionalEventStreamWriterImpl<Type> implements TransactionalEventStreamWriter<Type> {
+
+    private final Stream stream;
+    private final Serializer<Type> serializer;
+    private final SegmentOutputStreamFactory outputStreamFactory;
+    private final Controller controller;
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+    private final EventWriterConfig config;
+    private final ExecutorService retransmitPool;
+    private final Pinger pinger;
+    
+    TransactionalEventStreamWriterImpl(Stream stream, Controller controller, SegmentOutputStreamFactory outputStreamFactory,
+            Serializer<Type> serializer, EventWriterConfig config, ExecutorService retransmitPool) {
+        this.stream = Preconditions.checkNotNull(stream);
+        this.controller = Preconditions.checkNotNull(controller);
+        this.outputStreamFactory = Preconditions.checkNotNull(outputStreamFactory);
+        this.serializer = Preconditions.checkNotNull(serializer);
+        this.config = config;
+        this.retransmitPool = Preconditions.checkNotNull(retransmitPool);
+        this.pinger = new Pinger(config, stream, controller);
+    }
+
+    private static class TransactionImpl<Type> implements Transaction<Type> {
+
+        private final Map<Segment, SegmentTransaction<Type>> inner;
+        private final UUID txId;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+        private final Controller controller;
+        private final Stream stream;
+        private final Pinger pinger;
+        private StreamSegments segments;
+
+        TransactionImpl(UUID txId, Map<Segment, SegmentTransaction<Type>> transactions, StreamSegments segments,
+                Controller controller, Stream stream, Pinger pinger) {
+            this.txId = txId;
+            this.inner = transactions;
+            this.segments = segments;
+            this.controller = controller;
+            this.stream = stream;
+            this.pinger = pinger;
+        }
+        
+        /**
+         * Create closed transaction
+         */
+        TransactionImpl(UUID txId, Controller controller, Stream stream) {
+            this.txId = txId;
+            this.inner = null;
+            this.segments = null;
+            this.controller = controller;
+            this.stream = stream;
+            this.pinger = null;
+            this.closed.set(true);
+        }
+
+        /**
+         * Uses the transactionId to generate the routing key so that we only need to use one segment.
```

Comment:
I'm not sure what the reasoning behind using a single segment is. How is that preferable over round-robin across all available segments?

Reply:
Round robin would result in the events going to different readers and being processed out of order.